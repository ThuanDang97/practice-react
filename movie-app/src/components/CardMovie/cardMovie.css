.card {
  max-width: 220px;
  position: relative;
  top: 0;
  left: 0;
  border: 1px solid #e3e3e3;
  border-radius: 8px;
  overflow: hidden;
  margin-top: 30px;
  box-shadow: rgb(0 0 0 / 24%) 0px 3px 8px;
}

.card .content {
  width: 100%;
  padding: 26px 10px 12px 10px;
  position: relative;
  white-space: normal;
  display: flex;
  align-content: flex-start;
  flex-wrap: wrap;
  align-items: flex-start;
}

.hover {
  visibility: hidden;
  position: absolute;
  top: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.5);
  -webkit-backdrop-filter: blur(20px);
  backdrop-filter: blur(20px);
  width: 100%;
  height: 100%;
  border-radius: 8px;
  z-index: 5;
  transition: linear 0.1s;
  opacity: 0;
  display: flex;
  justify-content: center;
}

.on {
  visibility: visible;
  opacity: 1;
}

.option_content {
  align-self: center;
  padding: 2px;
  background: white;
  border-radius: 5px;
}

.option_content :hover {
  background: rgb(3, 37, 65);
}
