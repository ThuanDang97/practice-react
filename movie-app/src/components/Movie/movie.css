.movie {
  width: 100%;
  max-height: 600px;
}

.banner {
  max-width: 1440px;
  display: flex;
  margin: 0 auto;
}

.banner .banner_image {
  position: absolute;
  right: 0;
  height: 600px;
  width: 1520px;
}

.banner .layer {
  width: 100%;
  height: 600px;
  position: absolute;
  right: 0;
  background-image: linear-gradient(
    to right,
    rgba(0, 0, 0, 1) 150px,
    rgba(220.5, 220.5, 220.5, 0.84) 100%
  );
}

.banner .content {
  z-index: 1;
  padding: 75px 40px;
}
