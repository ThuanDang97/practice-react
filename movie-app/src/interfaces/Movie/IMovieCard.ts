import { IMovie } from './IMovie';

export interface IMovieCard extends IMovie {
  is_favorite: boolean;
}
