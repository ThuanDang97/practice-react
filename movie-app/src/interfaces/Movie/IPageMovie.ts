import { IMovie } from './IMovie';

export interface IPageMovie {
  page: number;
  results: IMovie[];
  total_pages: number;
  total_results: number;
}
