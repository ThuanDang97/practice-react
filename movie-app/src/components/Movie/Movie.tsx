import { IMovieInfo } from '../../interfaces/Movie/IMovieInfo';
import ContentMovie from './ContentMovie/ContentMovie';
import './movie.css';

type TProps = {
  movie: IMovieInfo;
};

const Movie = (props: TProps): JSX.Element => {
  const { movie } = props;

  return (
    <section className="movie">
      <div className="banner">
        <img
          src={`https://www.themoviedb.org/t/p/w1920_and_h800_multi_faces${movie?.backdrop_path}`}
          alt={movie?.original_title}
          className="banner_image"
        />
        <span className="layer"></span>
        <ContentMovie
          posterPath={movie?.poster_path!}
          homepage={movie?.homepage!}
          title={movie?.title!}
          releaseDate={movie?.release_date!}
          genres={movie?.genres!}
          runtime={movie?.runtime!}
          overview={movie?.overview!}
          scoreVoteAverage={movie?.vote_average!}
        />
      </div>
    </section>
  );
};

export default Movie;
