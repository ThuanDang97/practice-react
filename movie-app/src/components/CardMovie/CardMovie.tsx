import { memo, MouseEvent, useCallback, useState } from 'react';
import { optionsContent } from '../../constants/MockData';
import { IMovieCard } from '../../interfaces/Movie/IMovieCard';
import './cardMovie.css';
import ContentCard from './ContentCard/ContentCard';
import ImageCard from './ImageCard/ImageCard';
import Popup from '../Popup/Popup';

type TProps = {
  movie: IMovieCard;
  onUpdateStatus: (e: MouseEvent<HTMLDivElement>) => void;
};

const CardMovie = (props: TProps): JSX.Element => {
  const { movie, onUpdateStatus } = props;

  const [selectedCard, setSelectedCard] = useState(0);

  const handleHover = useCallback((id: number) => {
    setSelectedCard(id);
  }, []);

  const removeActions = () => {
    setSelectedCard(0);
  };

  const renderActions = (value: boolean) => {
    return (
      <div className="hover on" onClick={removeActions}>
        <div className="option_content">
          {optionsContent.map((option) => (
            <Popup
              key={option.id}
              src={option.src}
              alt={option.alt}
              title={option.title}
              color={value && option.title === 'Favorite' ? 'pink' : 'black'}
              onHandleClick={onUpdateStatus}
            />
          ))}
        </div>
      </div>
    );
  };

  return (
    <>
      <div className="card">
        <ImageCard
          id={movie.id!}
          originalTitle={movie.original_title!}
          posterPath={movie.poster_path!}
          onHandleHover={handleHover}
        />
        <ContentCard
          id={movie.id!}
          title={movie.title!}
          voteAverage={movie?.vote_average!}
          releaseDate={movie.release_date!}
        />
        {selectedCard === movie.id && renderActions(movie?.is_favorite)}
      </div>
    </>
  );
};

export default memo(CardMovie);
