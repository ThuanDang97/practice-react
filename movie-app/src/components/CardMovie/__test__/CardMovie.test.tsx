import { screen, render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { IMovieCard } from '../../../interfaces/Movie/IMovieCard';
import { setup } from '../../../support/Setup';
import CardMovie, { TProps } from '../CardMovie';

describe('Test Card Movie component', () => {
  const mockData: IMovieCard = {
    adult: false,
    backdrop_path: '/wcKFYIiVDvRURrzglV9kGu7fpfY.jpg',
    genre_ids: [],
    id: 453395,
    is_favorite: true,
    original_language: 'en',
    original_title: 'Doctor Strange in the Multiverse of Madness',
    overview:
      'Doctor Strange, with the help of mystical allies both old and new, traverses the mind-bending and dangerous alternate realities of the Multiverse to confront a mysterious new adversary.',
    popularity: 7647.02,
    poster_path: '/9Gtg2DzBhmYamXBS1hKAhiwbBKS.jpg',
    release_date: '2022-05-04',
    title: 'Doctor Strange in the Multiverse of Madness',
    video: false,
    vote_average: 7.6,
    vote_count: 2915,
  };

  const defaultProps: TProps = {
    movie: mockData,
    onUpdateStatus: jest.fn(),
  };

  const component = (
    <Router>
      <CardMovie {...defaultProps} />;
    </Router>
  );

  test('Component Card Movie matches DOM Snapshot', () => {
    const { container } = render(component);

    expect(container).toMatchSnapshot();
  });

  test('Component Card Movie should called removeActions', async () => {
    const { user } = setup(component);

    await user.click(screen.getByTestId('optional'));
    await user.click(screen.getByTestId('action'));
  });
});
