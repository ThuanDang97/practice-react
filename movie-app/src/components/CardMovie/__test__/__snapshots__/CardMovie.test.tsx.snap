// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Card Movie component Component Card Movie matches DOM Snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="image"
    >
      <div
        class="wrapper"
      >
        <a
          href="/movie/453395"
        >
          <img
            alt="Doctor Strange in the Multiverse of Madness"
            class="poster"
            src="https://www.themoviedb.org/t/p/w220_and_h330_face//9Gtg2DzBhmYamXBS1hKAhiwbBKS.jpg"
          />
        </a>
        <div
          class="option"
          data-testid="optional"
        >
          <span />
        </div>
      </div>
    </div>
    <div
      class="content"
    >
      <div
        class="consensus tight"
      >
        <div
          class="outer_ring"
        >
          <div
            class="user_score_chart"
            data-bar-color="#21d07a"
            data-percent="78.0"
            data-track-color="#204529"
          >
            <div
              class="percent"
            >
              <span
                class="icon"
                data-testid="score"
              >
                76
              </span>
            </div>
          </div>
        </div>
      </div>
      <h2>
        <a
          href="/movie/453395"
          title="Doctor Strange in the Multiverse of Madness"
        >
          Doctor Strange in the Multiverse of Madness
        </a>
      </h2>
      <p
        data-testid="release-date"
      >
        2022-05-04
      </p>
    </div>
  </div>
  ;
</div>
`;
